src/compact_cow_str.rs

```diff
@@ -0,0 +1,243 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::borrow::{Borrow, Cow};
+use std::cmp;
+use std::fmt;
+use std::hash;
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::Deref;
+use std::slice;
+use std::str;
+
+// All bits set except the highest
+const MAX_LEN: usize = !0 >> 1;
+
+// Only the highest bit
+const TAG: usize = MAX_LEN + 1;
+
+/// Like `Cow<'a, str>`, but with smaller `std::mem::size_of`. (Two words instead of four.)
+pub struct CompactCowStr<'a> {
+    // `tagged_len` is a tag in its highest bit, and the string length in the rest of the bits.
+    //
+    // * If the tag is 1, the memory pointed to by `ptr` is owned
+    //   and the lifetime parameter is irrelevant.
+    //   `ptr` and `len` are the components of a `Box<str>`.
+    //
+    // * If the tag is 0, the memory is borrowed.
+    //   `ptr` and `len` are the components of a `&'a str`.
+
+    ptr: *const u8,
+    tagged_len: usize,
+    phantom: PhantomData<&'a str>,
+}
+
+impl<'a> From<&'a str> for CompactCowStr<'a> {
+    #[inline]
+    fn from(s: &'a str) -> Self {
+        let len = s.len();
+        assert!(len <= MAX_LEN);
```

Comment:
I don’t know if any bit on the pointer is guaranteed to be available. Maybe not high bits since it seems like a userland process can get up to 3 GB on 32-bit. Not low bits since a subslice of `&str` can start at any byte.